Dockerfile.e2e: copy test sources

kolyshkin opened the pull request:
(this is a quick followup to #39116)

Package "gotest.tools/assert" uses source introspection to
print more info in case of assertion failure. When source code
is not available, it prints an error instead.

In other words, before this commit:

> --- SKIP: TestCgroupDriverSystemdMemoryLimit (0.00s)
>     cgroupdriver_systemd_test.go:32: failed to parse source file: /go/src/github.com/docker/docker/integration/system/cgroupdriver_systemd_test.go: open /go/src/github.com/docker/docker/integration/system/cgroupdriver_systemd_test.go: no such file or directory
>     cgroupdriver_systemd_test.go:32:

and after:

> --- SKIP: TestCgroupDriverSystemdMemoryLimit (0.09s)
>    cgroupdriver_systemd_test.go:32: !hasSystemd()

This increases the resulting image size by about 2 MB
on my system (from 758 to 760 MB).

kolyshkin committed:
Dockerfile.e2e: copy test sources

Package "gotest.tools/assert" uses source introspection to
print more info in case of assertion failure. When source code
is not available, it prints an error instead.

In other words, before this commit:

> --- SKIP: TestCgroupDriverSystemdMemoryLimit (0.00s)
>     cgroupdriver_systemd_test.go:32: failed to parse source file: /go/src/github.com/docker/docker/integration/system/cgroupdriver_systemd_test.go: open /go/src/github.com/docker/docker/integration/system/cgroupdriver_systemd_test.go: no such file or directory
>     cgroupdriver_systemd_test.go:32:

and after:

> --- SKIP: TestCgroupDriverSystemdMemoryLimit (0.09s)
>    cgroupdriver_systemd_test.go:32: !hasSystemd()

This increases the resulting image size by about 2 MB
on my system (from 758 to 760 MB).

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented on Dockerfile.e2e:
```diff
@@ -43,8 +43,11 @@ FROM base AS builder
 ENV DOCKER_GITCOMMIT=${DOCKER_GITCOMMIT:-undefined}
 COPY . .
+# Copy test sources tests that use assert can print errors
+RUN mkdir -p /build/${PWD} && find integration integration-cli -name \*_test.go -exec cp --parents '{}' /build${PWD} \;
```
I'm using $PWD here as it points to /go/src/github.com/docker/docker and this is where binaries expect the files to be in. This is also why the last COPY statement (below) has destination as /

thaJeztah reviewed (changes requested):
Left some comments, but otherwise looks good 🤗

kolyshkin committed:
Dockerfile.e2e: fix DOCKER_GITCOMMIT handling

1. There is no need to persist DOCKER_GITCOMMIT,
as it's not needed for runtime, only for build.
So, remove ENV.

2. In case $GITCOMMIT is not defined during build time
(and it happens if .git directory is not present),
we still need to have some value set, so set it to
`undefined`. Otherwise we'll have something like

>  => ERROR [builder 2/3] RUN hack/make.sh build-integration-test-binary
> ------
>  > [builder 2/3] RUN hack/make.sh build-integration-test-binary:
> #32 0.488
> #32 0.505 error: .git directory missing and DOCKER_GITCOMMIT not specified
> #32 0.505   Please either build with the .git directory accessible, or specify the
> #32 0.505   exact (--short) commit hash you are building using DOCKER_GITCOMMIT for
> #32 0.505   future accountability in diagnosing build issues.  Thanks!

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented on Dockerfile.e2e:
```diff
@@ -43,8 +43,11 @@ FROM base AS builder
 ARG DOCKER_GITCOMMIT
```
DONE

kolyshkin commented:
@seemethere PTAL (second patch)

thaJeztah commented on Dockerfile.e2e:
```diff
@@ -40,11 +40,13 @@ RUN CGO_ENABLED=0 go build -buildmode=pie -o /build/httpserver/httpserver github
+# Copy test sources tests that use assert can print errors
+RUN mkdir -p /build${PWD} && find integration integration-cli -name \*_test.go -exec cp --parents '{}' /build${PWD} \;
+# Build and install test binaries
+ARG DOCKER_GITCOMMIT=${DOCKER_GITCOMMIT:-undefined}
```
This should do the job;

```suggestion
ARG DOCKER_GITCOMMIT=undefined
```

kolyshkin replied:
Right; fixed

thaJeztah reviewed (approved):
LGTM, thanks!
